### DISCUSSION TITLE
Protein inheritance

### DISCUSSION THREAD

**Discussion by Yuri05** [Open-Systems-Pharmacology/QualificationPlan#161] [Ideas]
At the moment only complete building blocks (and for simulations: simulation parameter values) can be inherited from another project within a qualification plan: 
![grafik](https://user-images.githubusercontent.com/25061876/112809771-a6931900-907a-11eb-8864-5ac72e86cb2b.png)

However for Individuals/Populations it might be required to onherit not the **whole building block** but **only the proteins**.
For this, qualification plan could be extended with the **Protein Inheritance** feature, e.g. something like this:
![grafik](https://user-images.githubusercontent.com/25061876/112811340-40a79100-907c-11eb-85e6-6217088ab16a.png)

Then Protein (expressions) of a building block would be replaced by the protein expressions of the parent.

Question / Topics to discuss:
1) Should proteins always be **replaced** or should it be possible also to **merge**? E.g. if the parent-BB contains proteins E1, E2 and the child-BB contains proteins E3, E4: would the result always contain only E1, E2 or E1, E2, E3, E4?

2) Should protein inheritance be defined only for **individuals**? Or also for **populations**?

3) For proteins previously **not available** in a simulation: how should we proceed with the simulation mapping `Compound-Process <=> Individual Protein` ? **In my eyes is this the most critical point of the whole discussion**

4) If previously available proteins are **deleted** by the inheritance: how should we proceed if e.g. curves related to a deleted protein were referenced in the project (e.g. plotted/used in a PI/...)?

**Comment by Yuri05** [comment]
>1) Should proteins always be **replaced** or should it be possible also to **merge**? E.g. if the parent-BB contains proteins E1, E2 and the child-BB contains proteins E3, E4: would the result always contain only E1, E2 or E1, E2, E3, E4?

I would vote for "replace only". No merge.

**Comment by Yuri05** [comment]
>3) For proteins previously **not available** in a simulation: how should we proceed with the simulation mapping `Compound-Process <=> Individual Protein` ? **In my eyes is this the most critical point of the whole discussion**

Many possible solutions here and a lot of space for side effects. 
If Individual (or population) in the qualification plan inherits its proteins, then all simulations where this individual/population is used should become an info "add default mappings for this particular individual/population" or something in this direction. Thus:
 - if a previously available protein was removed from the individual/population due to inheritance: its mappings must be removed from the simulation configuration
 - if a new protein was added: the default mappings for this protein would be used in the simulation (**which is pretty error prone and migth be NOT what the user has in mind!** e.g. right now the default mapping takes the compound process which was created *first*)
<img width="250" src="https://user-images.githubusercontent.com/25061876/112816254-7f8c1580-9081-11eb-8e6a-9c4edc448b1b.png"> ➡️ <img width="400" src="https://user-images.githubusercontent.com/25061876/112815968-29b76d80-9081-11eb-8f63-8b60edbe0823.png">

Thus either something like *default flag* must be added to compound process definition.
Or the mapping must be defined **in the qualification plan** (which I would NOT do for sure!)

Even if something like "default process" would be defined in a compound, there are still situations where the default mapping would not create everything what the user wants (e.g. in case where **several metabolites** are created by the same process; some DDI combinations, etc.)

To be honest, the whole procedure is so complicated and error-prone...

**Comment by PavelBal** [comment]
1) I definitely vote against deleting proteins!

I do not fully understand the problem with adding new proteins. If protein `A` is added by the inheritance to a simulation, then the configuration of such a simulation does not has a mapping of any compound to the protein, right?

**Comment by Yuri05** [comment]
>If protein A is added by the inheritance to a simulation, then the configuration of such a simulation does not has a mapping of any compound to the protein, right?

Right. If we keep it like this (new proteins are just added to the building block but not used in simulations) - much easier to implement. However: what is then the benefit of the feature? Just the possibility to **update** the properties (relExp etc.) of **previously existing** proteins?

>I definitely vote against deleting proteins!

This would solve some of the problems for sure :)
